services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java

```diff
@@ -0,0 +1,102 @@
+package org.keycloak.authentication.authenticators.sessionlimits;
+
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.models.AuthenticatorConfigModel;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.services.managers.AuthenticationManager;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.ws.rs.core.Response;
+import org.keycloak.events.Errors;
+import org.keycloak.services.messages.Messages;
+
+public class UserSessionLimitsAuthenticator extends AbstractSessionLimitsAuthenticator {
+
+    private static Logger logger = Logger.getLogger(UserSessionLimitsAuthenticator.class);
+
+    String behavior;
+
+    public UserSessionLimitsAuthenticator(KeycloakSession session) {
+        this.session = session;
+    }
+
+    @Override
+    public void authenticate(AuthenticationFlowContext context) {
+        AuthenticatorConfigModel authenticatorConfig = context.getAuthenticatorConfig();
+        Map<String, String> config = authenticatorConfig.getConfig();
+
+        // Get the configuration for this authenticator
+        behavior = config.get(UserSessionLimitsAuthenticatorFactory.BEHAVIOR);
+        int userRealmLimit = getIntConfigProperty(UserSessionLimitsAuthenticatorFactory.USER_REALM_LIMIT, config);
+        int userClientLimit = getIntConfigProperty(UserSessionLimitsAuthenticatorFactory.USER_CLIENT_LIMIT, config);
+
+        if (context.getRealm() != null && context.getUser() != null) {
+
+            // Get the session count in this realm for this specific user
+            List<UserSessionModel> userSessionsForRealm = session.sessions().getUserSessions(context.getRealm(), context.getUser());
+            int userSessionCountForRealm = userSessionsForRealm.size();
+
+            // Get the session count related to the current client for this user
+            ClientModel currentClient = context.getAuthenticationSession().getClient();
+            logger.debugf("session-limiter's current keycloak clientId: %s", currentClient.getClientId());
+
+            List<UserSessionModel> userSessionsForClient = userSessionsForRealm.stream().filter(session -> session.getAuthenticatedClientSessionByClient(currentClient.getId()) != null).collect(Collectors.toList());
```

Comment:
Additional point: When the count of clients per user ( USER_CLIENT_LIMIT ) is disabled (which IMO should be default behaviour as IMO most common use-case is limit of total user sessions per user), it will be nice to make the code in a way that call to `session.getAuthenticatedClientSessionByClient` is avoided as this has some additional performance penalty.

Reply:
Yes, good point! It is now called conditionally.